In [1]:
class TrieNode:
    def __init__(self):
        self.filhos = {}
        self.fim_palavra = False


class Trie:
    def __init__(self):
        self.raiz = TrieNode()

    # INSERIR
    def inserir(self, palavra):
        atual = self.raiz

        for letra in palavra.lower():
            if letra not in atual.filhos:
                atual.filhos[letra] = TrieNode()

            atual = atual.filhos[letra]

        atual.fim_palavra = True

    # BUSCAR
    def buscar(self, palavra):
        atual = self.raiz

        for letra in palavra.lower():
            if letra not in atual.filhos:
                return False

            atual = atual.filhos[letra]

        return atual.fim_palavra

    # REMOVER
    def remover(self, palavra):

        def remover_rec(no, palavra, indice):

            if indice == len(palavra):

                if not no.fim_palavra:
                    return False

                no.fim_palavra = False

                return len(no.filhos) == 0

            letra = palavra[indice]

            if letra not in no.filhos:
                return False

            apagar = remover_rec(
                no.filhos[letra],
                palavra,
                indice + 1
            )

            if apagar:
                del no.filhos[letra]

                return (
                    len(no.filhos) == 0
                    and not no.fim_palavra
                )

            return False

        remover_rec(self.raiz, palavra.lower(), 0)

    # LISTAR
    def listar(self):

        palavras = []

        def dfs(no, prefixo):

            if no.fim_palavra:
                palavras.append(prefixo)

            for letra, filho in no.filhos.items():
                dfs(filho, prefixo + letra)

        dfs(self.raiz, "")

        return palavras

    # AUTOCOMPLETE
    def autocomplete(self, prefixo):

        atual = self.raiz

        for letra in prefixo.lower():

            if letra not in atual.filhos:
                return []

            atual = atual.filhos[letra]

        resultados = []

        def dfs(no, palavra):

            if no.fim_palavra:
                resultados.append(palavra)

            for letra, filho in no.filhos.items():
                dfs(filho, palavra + letra)

        dfs(atual, prefixo.lower())

        return resultados

    # DISTÂNCIA DE EDIÇÃO 
    def distancia(self, a, b):

        linhas = len(a) + 1
        colunas = len(b) + 1

        matriz = [
            [0 for _ in range(colunas)]
            for _ in range(linhas)
        ]

        for i in range(linhas):
            matriz[i][0] = i

        for j in range(colunas):
            matriz[0][j] = j

        for i in range(1, linhas):

            for j in range(1, colunas):

                custo = 0

                if a[i - 1] != b[j - 1]:
                    custo = 1

                matriz[i][j] = min(
                    matriz[i - 1][j] + 1,
                    matriz[i][j - 1] + 1,
                    matriz[i - 1][j - 1] + custo
                )

        return matriz[-1][-1]

    # AUTOCORREÇÃO
    def autocorrecao(self, palavra):

        palavras = self.listar()

        melhor = None
        menor_distancia = float("inf")

        for p in palavras:

            d = self.distancia(
                palavra.lower(),
                p.lower()
            )

            if d < menor_distancia:
                menor_distancia = d
                melhor = p

        return melhor


# PALAVRAS MAIS COMUNS
palavras = [
    "a","o","e","de","do","da","em","um","uma","para",
    "com","não","que","se","por","mais","como","mas","foi","ao",
    "ele","ela","dos","das","tem","já","ou","ser","sua","seu",
    "tambem","ha","quando","muito","nos","nas","pela","pelo","entre","era",
    "sobre","isso","sem","mesmo","ainda","ate","anos","tempo","onde","porque",
    "quem","todo","toda","todos","todas","cada","fazer","vida","dia","dias",
    "homem","mulher","casa","cidade","estado","pais","brasil","trabalho","empresa","dinheiro",
    "escola","professor","aluno","livro","computador","internet","sistema","processo","programa","projeto",
    "dados","arquivo","rede","usuario","senha","seguranca","tecnologia","software","hardware","codigo",
    "python","java","estrutura","algoritmo","lista","pilha","fila","arvore","grafo","banco",
    "informacao","desenvolvimento","aplicacao","servidor","cliente","teste","documento","arquivo"
]

trie = Trie()

for palavra in palavras:
    trie.inserir(palavra)


print("*" * 60)
print("TESTE DE BUSCA")
print("*" * 60)

print("casa:", trie.buscar("casa"))
print("python:", trie.buscar("python"))
print("banana:", trie.buscar("banana"))


print("\n" + "*" * 60)
print("TESTE DE REMOÇÃO")
print("*" * 60)

print("Antes:", trie.buscar("casa"))

trie.remover("casa")

print("Depois:", trie.buscar("casa"))


print("\n" + "*" * 60)
print("TESTE DE LISTAGEM")
print("*" * 60)

lista_palavras = trie.listar()

print("Quantidade de palavras:", len(lista_palavras))

print(lista_palavras[:20])


print("\n" + "*" * 60)
print("TESTE DE AUTOCOMPLETE")
print("*" * 60)

print(trie.autocomplete("pro"))


print("\n" + "*" * 60)
print("TESTE DE AUTOCORREÇÃO")
print("*" * 60)

print("progama -", trie.autocorrecao("progama"))
print("estrutra -", trie.autocorrecao("estrutra"))
print("tecnolojia -", trie.autocorrecao("tecnolojia"))

************************************************************
TESTE DE BUSCA
************************************************************
casa: True
python: True
banana: False

************************************************************
TESTE DE REMOÇÃO
************************************************************
Antes: True
Depois: False

************************************************************
TESTE DE LISTAGEM
************************************************************
Quantidade de palavras: 106
['a', 'ao', 'ainda', 'ate', 'anos', 'aluno', 'algoritmo', 'arquivo', 'arvore', 'aplicacao', 'o', 'ou', 'onde', 'e', 'em', 'empresa', 'ele', 'ela', 'entre', 'era']

************************************************************
TESTE DE AUTOCOMPLETE
************************************************************
['professor', 'processo', 'programa', 'projeto']

************************************************************
TESTE DE AUTOCORREÇÃO
*************************************************

***Resultados dos Testes***

Os testes realizados demonstraram o correto funcionamento da Trie. Na busca, as palavras "casa" e "python" foram encontradas, enquanto "banana" não estava cadastrada. O teste de remoção confirmou que a palavra "casa" foi excluída com sucesso da estrutura. A função de listagem retornou 106 palavras, mostrando que todas as palavras armazenadas puderam ser recuperadas. No teste de autocompletar, o prefixo "pro" retornou as palavras professor, processo, programa e projeto. Já a autocorreção sugeriu corretamente programa para "progama", estrutura para "estrutra" e tecnologia para "tecnolojia", comprovando a eficiência do algoritmo de correção.